### PR TITLE
rsp.inc: unify support for e(x) syntax

### PR DESCRIPTION
**Pull request by snacchus**
The vector element syntax that was introduced for computational vector opcodes is now supported in `mtc2`/`mfc2` and vector load/store opcodes as well. Until now, you had to specify the element as the byte offset into the vector register for those opcodes. Now it is possible to specify the lane index directly instead. For example:
```
# Before:
lsv $v00,4,  0x10,s4
mfc2 t0, $v01,6

# After:
lsv $v00,e(2),  0x10,s4
mfc2 t0, $v01,e(3)
```

### CODE REVIEW

File: src/audio/rsp_mixer.S [hunk 1]
```diff
@@ -276,9 +276,9 @@ command_exec:
 	#define samples_left    t4
 	#define outptr          s8
 
-	vxor v_zero, v_zero, v_zero,0
+	vxor v_zero, v_zero, v_zero
 	li t0, %lo(VCONST_1)
-	lqv v_const1,0, 0,t0
+	lqv v_const1,e(0), 0,t0
```
Comment (rasky) [inline src/audio/rsp_mixer.S]: Honestly I don’t know if I’d make these changes. It makes the code more verbose and honestly the real solution here would be making that argument optional. If that’s not easy to do, maybe just leave 0?

File: include/rsp.inc [hunk 2]
```diff
@@ -474,52 +539,52 @@ makeOpInstruction vsubc, 0b010101
 makeOpInstruction vxor, 0b101100
 
 /** @brief Load Byte into Vector Register */
-makeLsInstructionByte load, lbv, 0b00000
+makeLsInstructionByte load, byte, lbv, 0b00000
```
Comment (rasky) [inline include/rsp.inc:954]: I’ll review these properly

File: src/audio/rsp_mixer.S (hunk 1 above)
Comment (snacchus) [reply src/audio/rsp_mixer.S]: Yeah you're right. I'll look into this again. I wonder if there is a good, general approach to providing multiple "overloads" of GAS macros

**Comment by snacchus** [comment]
Update: In addition to the `e(x)` syntax, all relevant opcodes now also support a new "dot syntax":
```
vadd $v01, $v02, $v03.h2   # equivalent to $v03,e(2h)
vsub $v04, $v05.e4         # equivalent to $v05,e(4)

lsv $v10.e5, 0x10,s0      # equivalent to $v10,10
ssv $v11.b4, 0x00,s1      # equivalent to $v11,4

mfc2 t0, $v20.e7           # equivalent to $v20,14
```

Also, the element is now optional everywhere. For computational opcodes, it defaults to the normal vector access (like before). For all others, the default is lane 0 (or byte offset 0).

File: src/audio/rsp_mixer.S (hunk 1 above)
Comment (snacchus) [reply src/audio/rsp_mixer.S]: The element is now optional

File: include/rsp.inc [hunk 3]
```diff
@@ -473,138 +732,229 @@ makeOpInstruction vsubc, 0b010101
...
 /** @brief Load Double into Vector Register */
-makeLsInstructionDouble load, ldv, 0b00011
+makeLsInstructionDouble load, byte, ldv, 0b00011
 /** @brief Load Packed Fourth into Vector Register */
-makeLsInstructionQuad load, lfv, 0b01001
+makeLsInstructionQuad load, lane, lfv, 0b01001
 /** @brief Load Packed Half into Vector Register */
-makeLsInstructionQuad load, lhv, 0b01000
+makeLsInstructionQuad load, lane, lhv, 0b01000
```
Comment (rasky) [inline include/rsp.inc]: Both lhv and lfv do accept byte offsets.

File: include/rsp.inc [hunk 4]
```diff
@@ -473,138 +732,229 @@ makeOpInstruction vsubc, 0b010101
...
+makeLsInstructionQuad load, byte, lqv, 0b00100
 /** @brief Load Quad (Rest) into Vector Register */
-makeLsInstructionQuad load, lrv, 0b00101
+makeLsInstructionQuad load, byte, lrv, 0b00101
 /** @brief Load Short into Vector Register */
-makeLsInstructionShort load, lsv, 0b00001
+makeLsInstructionShort load, byte, lsv, 0b00001
 /** @brief Load Transpose into Vector Register */
-makeLsInstructionQuad load, ltv, 0b01011
+makeLsInstructionQuad load, byte, ltv, 0b01011
```
Comment (rasky) [inline include/rsp.inc]: ltv and stv are an exception :)

In theory the element is treated as a byte offset, but in practice the lowest bit is ignored. So semantically speaking I would say that it should only accept a lane, but encoding-wise a byte offset should be written.

File: include/rsp.inc [hunk 5]
```diff
@@ -473,138 +732,229 @@ makeOpInstruction vsubc, 0b010101
...
+makeLsInstructionByte store, byte, sbv, 0b00000
 /** @brief Store Double from Vector Register */
-makeLsInstructionDouble store, sdv, 0b00011
+makeLsInstructionDouble store, byte, sdv, 0b00011
 /** @brief Store Packed Fourth from Vector Register */
-makeLsInstructionQuad store, sfv, 0b01001
+makeLsInstructionQuad store, lane, sfv, 0b01001
 /** @brief Store Packed Half from Vector Register */
-makeLsInstructionQuad store, shv, 0b01000
+makeLsInstructionQuad store, lane, shv, 0b01000
```
Comment (rasky) [inline include/rsp.inc]: Both sfv and shv accept a byte offset.

File: include/rsp.inc (hunk 3 above)
Comment (snacchus) [reply include/rsp.inc]: fixed

File: include/rsp.inc (hunk 4 above)
Comment (snacchus) [reply include/rsp.inc]: I've added a special encoding for ltv and stv

File: include/rsp.inc (hunk 5 above)
Comment (snacchus) [reply include/rsp.inc]: fixed

File: include/rsp.inc [hunk 6]
```diff
@@ -464,147 +770,270 @@ makeOpInstruction vrsq, 0b110100
...
+    .endif
+
+    .ifnb \vs
+        # \vs is element
+        vsarInternal \vd, \vd, \vd, \vs
+        .exitm
+    .endif
+
+    .error "Element is required"
+.endm
```
Comment (rasky) [inline include/rsp.inc:811]: `vsar` basically only cares about `vd` and `elem`. `vt` and `vs` are ignored, to the best of my understanding.

So the only good syntax should be 2-arg. Given that we've forced 4-args up until now, I would probably support that as backward compatibility as well, marking it as deprecated in comments. I would not introduce the 3-arg syntax since it's not required.

Also I'd change `vsarInternal` to use the 2-arg syntax, so that it's clear that the other elements are ignored (can be forced to 0 while building the opcode).

File: .github/workflows/build-documentation.yml [hunk 7]
```diff
@@ -26,7 +26,7 @@ jobs:
           echo "::set-output name=default_ref::${{ format('refs/heads/{0}', github.event.repository.default_branch) }}"
 
       - name: Run Doxygen
-        uses: mattnotmitt/doxygen-action@v1
+        uses: mattnotmitt/doxygen-action@master
```
Comment (meeq) [inline .github/workflows/build-documentation.yml]: This change seems orthogonal to the rest of the PR, and is not good practice: "master" is not a tagged version, so it can cause breakage without warning and provides no guarantees of compatibility.

It looks like `1.9.2` is the latest stable version of the package. `1.9.3` is not considered stable because it depends on an `edge` version of Alpine Linux.

Comment (snacchus) [reply .github/workflows/build-documentation.yml]: The reason for this was the doxygen build being broken at the time, which made the CI build fail. Rasky and I then came up with this solution. That issue seems resolved now though (some other PRs made it through without this change in the meantime) so I'll roll it back.
